core/state_transition.go

```diff
@@ -187,8 +187,14 @@ func (st *StateTransition) to() common.Address {
 }
 
 func (st *StateTransition) buyGas() error {
-	mgval := new(big.Int).Mul(new(big.Int).SetUint64(st.msg.Gas()), st.gasPrice)
-	if have, want := st.state.GetBalance(st.msg.From()), mgval; have.Cmp(want) < 0 {
+	mgval := new(big.Int).SetUint64(st.msg.Gas())
+	mgval = mgval.Mul(mgval, st.gasPrice)
+	balanceCheck := mgval
+	if st.feeCap != nil {
```

Comment:
In theory this `feeCap` should always be non-nil, since for all legacy transactions they just return the `GasPrice` instead.

Perhaps we can define a `GasCost` function in the `types` package to handle the complexity.

But anyway, this code is also fine, but just with some additional calculation.